contracts/cosmwasm-vm/cw-integration/tests/test_xcall.rs

```diff
@@ -0,0 +1,360 @@
+            revision_number: 0,
+            revision_height: 12345,
+        }),
+        timeout_timestamp: 17747483838282,
+    };
+    let packet_bytes = hex::encode(packet.encode_to_vec());
+    ctx.app.execute_contract(
+        ctx.sender.clone(),
+        ctx.get_ibc_core(),
+        &cw_mock_ibc_core::msg::ExecuteMsg::ReceivePacket {
+            message: packet_bytes,
+        },
+        &[],
+    )
+}
+
+pub fn call_register_connection(ctx: &mut TestContext) -> Result<AppResponse, AppError> {
+    ctx.app.execute_contract(
+        ctx.sender.clone(),
+        ctx.get_ibc_core(),
+        &cw_mock_ibc_core::msg::ExecuteMsg::RegisterXcall {
+            address: ctx.get_xcall_ibc_connection(),
+        },
+        &[],
+    )
+}
+
+#[test]
+fn test_xcall_send_call_message() {
+    let mut ctx = setup_test();
+    call_set_xcall_host(&mut ctx).unwrap();
+    call_register_connection(&mut ctx).unwrap();
+    let src = ctx.get_xcall_ibc_connection().to_string();
+
+    let nid = "0x3.icon";
+    call_configure_connection(
+        &mut ctx,
+        "connection-1".to_string(),
+        nid.to_string(),
+        "client-1".to_string(),
+    )
+    .unwrap();
+    call_ibc_channel_connect(&mut ctx).unwrap();
+    let result = call_send_call_message(
+        &mut ctx,
+        &format!("{nid}/{MOCK_CONTRACT_TO_ADDR}"),
+        vec![src],
+        vec!["somedestination".to_string()],
+        vec![1, 2, 3],
+        None,
+    );
+    println!("{result:?}");
+    assert!(result.is_ok());
+    let result = result.unwrap();
+    let event = get_event(&result, "wasm-CallMessageSent").unwrap();
+    println!("{event:?}");
+    assert_eq!(
+        &format!("{nid}/{MOCK_CONTRACT_TO_ADDR}"),
+        event.get("to").unwrap()
+    );
+}
+
+#[test]
+fn test_xcall_send_call() {
+    let mut ctx = setup_test();
+    call_set_xcall_host(&mut ctx).unwrap();
+    call_register_connection(&mut ctx).unwrap();
+    let src = ctx.get_xcall_ibc_connection().to_string();
+    let dapp = ctx.get_dapp().to_string();
+
+    let nid = "0x3.icon";
+    call_configure_connection(
+        &mut ctx,
+        "connection-1".to_string(),
+        nid.to_string(),
+        "client-1".to_string(),
+    )
+    .unwrap();
+    call_ibc_channel_connect(&mut ctx).unwrap();
+    let message = AnyMessage::CallMessageWithRollback(CallMessageWithRollback {
+        data: vec![1, 2, 3],
+        rollback: "rollback-reply".as_bytes().to_vec(),
+    });
+    let envelope = Envelope::new(message, vec![src], vec!["somedestination".to_string()]);
+    let result = call_dapp_send_call(&mut ctx, format!("{nid}/{dapp}"), envelope);
+    println!("{result:?}");
+    assert!(result.is_ok());
+    let result = result.unwrap();
+    let event = get_event(&result, "wasm-CallMessageSent").unwrap();
+    println!("{event:?}");
+    assert_eq!(&format!("{nid}/{dapp}"), event.get("to").unwrap());
+}
+
+#[test]
+fn test_rollback_reply() {
```

Comment:
What does this test test? I having a hard time following it

Reply:
it is asserting that dapp can send reply when message with rollback is executed and we receive a message in ack with expected parameters.